src/transports/ws.rs

```diff
@@ -0,0 +1,258 @@
+//! WebSocket Transport
+
+extern crate websocket;
+
+use std::collections::BTreeMap;
+use std::sync::{atomic, Arc};
+
+use futures::sync::{mpsc, oneshot};
+use futures::{self, Future, Sink, Stream};
+use helpers;
+use parking_lot::Mutex;
+use rpc;
+use self::websocket::{ClientBuilder, OwnedMessage};
+use self::websocket::url::Url;
+use transports::Result;
+use transports::shared::{EventLoopHandle, Response};
+use transports::tokio_core::reactor;
+use {BatchTransport, Error, ErrorKind, RequestId, Transport};
+
+impl From<websocket::WebSocketError> for Error {
+  fn from(err: websocket::WebSocketError) -> Self {
+    ErrorKind::Transport(format!("{:?}", err)).into()
+  }
+}
+
+impl From<websocket::client::ParseError> for Error {
+  fn from(err: websocket::client::ParseError) -> Self {
+    ErrorKind::Transport(format!("{:?}", err)).into()
+  }
+}
+
+type Pending = oneshot::Sender<Result<Vec<Result<rpc::Value>>>>;
+
+/// A future representing pending WebSocket request, resolves to a response.
+pub type WsTask<F> = Response<F, Vec<Result<rpc::Value>>>;
+
+/// WebSocket transport
+#[derive(Debug, Clone)]
+pub struct WebSocket {
+  id: Arc<atomic::AtomicUsize>,
+  url: Url,
+  pending: Arc<Mutex<BTreeMap<RequestId, Pending>>>,
+  write_sender: mpsc::UnboundedSender<OwnedMessage>,
+}
+
+impl WebSocket {
+  /// Create new WebSocket transport with separate event loop.
+  /// NOTE: Dropping event loop handle will stop the transport layer!
+  pub fn new(url: &str) -> Result<(EventLoopHandle, Self)> {
+    let url = url.to_owned();
+    EventLoopHandle::spawn(move |handle| Self::with_event_loop(&url, &handle).map_err(Into::into))
+  }
+
+  /// Create new WebSocket transport within existing Event Loop.
+  pub fn with_event_loop(url: &str, handle: &reactor::Handle) -> Result<Self> {
+    trace!("Connecting to: {:?}", url);
+
+    let url: Url = url.parse()?;
+    let (write_sender, write_receiver) = mpsc::unbounded();
+    let pending: Arc<Mutex<BTreeMap<RequestId, Pending>>> = Arc::new(Mutex::new(BTreeMap::new()));
+
+    let ws_future = {
+      let pending_ = pending.clone();
+      let write_sender_ = write_sender.clone();
+
+      ClientBuilder::from_url(&url)
+        .add_protocol("web3.rs")
```

Comment:
Is the protocol required for anything in other clients? Parity will reject such requests, it expects protocol to be empty. Can we remove that?